### ISSUE TITLE
Prevent Re-initialization Error in VideoPlayerController in Event Listener

### ISSUE THREAD

**Issue by Mina-Zarif** [flutter/flutter#158433]
### Steps to reproduce

1. Use the VideoPlayerController to load a video.
2. Observe if eventListener() receives multiple VideoEventType.initialized events.

### Expected results

The VideoPlayerController should only initialize once, regardless of duplicate VideoEventType.initialized events.

### Actual results

When the VideoPlayerController receives multiple VideoEventType.initialized events, it attempts to re-initialize, which can lead to:

Errors: The initializingCompleter assertion fails, throwing an error indicating the controller has already been initialized.
Inconsistent State: The video player may behave unpredictably, as multiple initializations can interfere with playback controls, buffering, or seeking.
Performance Issues: Redundant initialization can cause unnecessary processing, affecting app performance.
The modified code eliminates these issues by preventing re-initialization, ensuring stable playback and consistent controller behavior.

### Code sample

<details open><summary>Code sample</summary>

```dart
void eventListener(VideoEvent event) {
  if (_isDisposed) return;

  // Prevent re-initialization if already initialized
  if (initializingCompleter.isCompleted && event.eventType == VideoEventType.initialized) {
    return;
  }

  switch (event.eventType) {
    case VideoEventType.initialized:
      value = value.copyWith(
        duration: event.duration,
        size: event.size,
        rotationCorrection: event.rotationCorrection,
        isInitialized: event.duration != null,
        errorDescription: null,
        isCompleted: false,
      );

      // Assert if initializingCompleter is already completed
      assert(
        !initializingCompleter.isCompleted,
        'VideoPlayerController already initialized. This is typically a '
        'sign that an implementation of the VideoPlayerPlatform '
        '(${_videoPlayerPlatform.runtimeType}) has a bug and is sending '
        'more than one initialized event per instance.',
      );

      // Complete initialization if not completed yet
      if (!initializingCompleter.isCompleted) {
        initializingCompleter.complete(null);
      }

      _applyLooping();
      _applyVolume();
      _applyPlayPause();
      break;

    case VideoEventType.completed:
      // Stop playback, set isPlaying to false, and move to the end position
      pause().then((void pauseResult) => seekTo(value.duration));
      value = value.copyWith(isCompleted: true);
      break;

    case VideoEventType.bufferingUpdate:
      value = value.copyWith(buffered: event.buffered);
      break;

    case VideoEventType.bufferingStart:
      value = value.copyWith(isBuffering: true);
      break;

    case VideoEventType.bufferingEnd:
      value = value.copyWith(isBuffering: false);
      break;

    case VideoEventType.isPlayingStateUpdate:
      value = value.copyWith(
        isPlaying: event.isPlaying,
        isCompleted: event.isPlaying == false ? value.isCompleted : false,
      );
      break;

    case VideoEventType.unknown:
      break;
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[[Upload media here]](https://github.com/user-attachments/assets/13809c60-2b36-4b0d-8600-ab41dd52a8c9)

</details>

### Logs

<details open><summary>Logs</summary>

```console

Unhandled Exception: PlatformException(VideoError, Video player had error androidx.media3.exoplayer.ExoPlaybackException: MediaCodecVideoRenderer error, index=0, format=Format(1, null, null, video/avc, avc1.42000A, -1, null, [720, 1600, 22.964258, ColorInfo(BT601, Limited range, Unset color transfer, false, 8bit Luma, 8bit Chroma)], [-1, -1]), format_supported=YES, null, null)
why that happened while initplayer 

```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[√] Flutter (Channel stable, 3.24.1, on Microsoft Windows [Version 10.0.22631.4391], locale en-US)
    • Flutter version 3.24.1 on channel stable at C:\src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5874a72aa4 (3 months ago), 2024-08-20 16:46:00 -0500
    • Engine revision c9b9d5780d
    • Dart version 3.5.1
    • DevTools version 2.37.2

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at C:\Users\Mina_\AppData\Local\Android\sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0--11609105)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Users\Mina_\AppData\Local\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Professional 2022 17.5.1)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Professional
    • Visual Studio Professional 2022 version 17.5.33424.131
    • Windows 10 SDK version 10.0.22000.0

[√] Android Studio (version 2024.1)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0--11609105)

[√] IntelliJ IDEA Ultimate Edition (version 2022.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2022.2.3
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin version 222.4345.14

[√] VS Code (version 1.95.2)
    • VS Code at C:\Users\Mina_\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.100.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22631.4391]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 130.0.6723.117
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 130.0.2849.80

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by Mina-Zarif** [comment]
report Re-initialization Error in VideoPlayerController

**Comment by huycozy** [comment]
You can check this https://github.com/flutter/flutter/issues/156047 to see if it assembles your case or not. 

Otherwise, for us to further investigate this issue, please provide a [completed and minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) that doesn’t include 3rd party plugins or complex production code. Thanks!

**Comment by Mina-Zarif** [comment]
Thank you for the response!

I reviewed Issue [#156047](https://github.com/flutter/flutter/issues/156047) and found that it describes similar symptoms, but it does not directly resolve the issue I encountered.

**Summary of My Solution**
After investigating further, I changed the code in the video player package itself, which solved the issue. Specifically, I added a check in the `eventListener()` method to ensure that re-initialization does not occur if the `VideoPlayerController` has already completed its initialization. This prevents the double-initialization issue that was causing errors.

**Code Changes**
Here’s a summary of the changes I made in `eventListener()`:


```
void eventListener(VideoEvent event) {
  if (_isDisposed) return;

  // Prevent re-initialization if already initialized
  if (initializingCompleter.isCompleted && event.eventType == VideoEventType.initialized) {
    return;
  }

  switch (event.eventType) {
    case VideoEventType.initialized:
      value = value.copyWith(
        duration: event.duration,
        size: event.size,
        rotationCorrection: event.rotationCorrection,
        isInitialized: event.duration != null,
        errorDescription: null,
        isCompleted: false,
      );
  // Complete initialization if not completed yet
      if (!initializingCompleter.isCompleted) {
        initializingCompleter.complete(null);
      }

      _applyLooping();
      _applyVolume();
      _applyPlayPause();
      break;

    // Other cases follow without changes...
  }
}
```

This adjustment checks whether `initializingCompleter` has already completed and prevents re-initialization if it has, stabilizing the video playback.

**Request**
Would it be possible to consider this check for a future update to the package? Alternatively, if this is intended behavior, please let me know if there’s a recommended way to handle this scenario without modifying the package code.

Thank you!

**Comment by huycozy** [comment]
It may work for your case (maybe just a workaround) but I'm not sure whether it is a proper fix for every case. It needs to pass reviews and tests, please check [Triage process for teams](https://github.com/flutter/flutter/blob/master/docs/triage/README.md#triage-process-for-teams).

Please provide a reproduction sample code so that I can try to reproduce the issue and forward this to an appropriate team. Thanks!

**Comment by Mina-Zarif** [comment]
Thank you for your help so far.

After testing with another project, I couldn’t reproduce the issue outside of my original project. This suggests that the duplicate initialization may be specific to the configuration or environment of that project rather than a general bug in the package. I’ll continue investigating my project’s setup to identify any factors that might contribute to this behavior, and I’ll update here if I find a root cause that could affect other users.


**Proposed Improvement to `eventListener()` for Performance and Stability**

However, during this investigation, I identified an optimization opportunity in the eventListener method that could benefit overall performance and stability. The current package implementation lacks a check to prevent redundant initialization when initializingCompleter is already completed. This could lead to unnecessary processing and potential errors if the initialized event is triggered multiple times.

**Package Source Code**
The current code in the package source looks like this:

```
void eventListener(VideoEvent event) {
  if (_isDisposed) {
    return;
  }

  switch (event.eventType) {
    case VideoEventType.initialized:
      value = value.copyWith(
        duration: event.duration,
        size: event.size,
        rotationCorrection: event.rotationCorrection,
        isInitialized: event.duration != null,
        errorDescription: null,
        isCompleted: false,
      );
      assert(
        !initializingCompleter.isCompleted,
        'VideoPlayerController already initialized. This is typically a '
        'sign that an implementation of the VideoPlayerPlatform '
        '(${_videoPlayerPlatform.runtimeType}) has a bug and is sending '
        'more than one initialized event per instance.',
      );
//  to end of fun
}
```

**Suggested Code Improvement**
In my project, I addressed this by adding a check to skip re-initialization if `initializingCompleter` is already completed:

```
void eventListener(VideoEvent event) {
  if (_isDisposed) {
    return;
  }

  // Prevent redundant re-initialization if already initialized
  if (initializingCompleter.isCompleted && event.eventType == VideoEventType.initialized) {
    return;
  }

  switch (event.eventType) {
    case VideoEventType.initialized:
      value = value.copyWith(
        duration: event.duration,
        size: event.size,
        rotationCorrection: event.rotationCorrection,
        isInitialized: event.duration != null,
        errorDescription: null,
        isCompleted: false,
      );
      assert(
        !initializingCompleter.isCompleted,
        'VideoPlayerController already initialized. This is typically a '
        'sign that an implementation of the VideoPlayerPlatform '
        '(${_videoPlayerPlatform.runtimeType}) has a bug and is sending '
        'more than one initialized event per instance.',
      );
//  to end of fun
}

```
**Benefits:**

**Improves Performance:** Reduces unnecessary processing when duplicate initialized events are triggered.

**Enhances Stability:** Prevents re-initialization, which could lead to errors or unexpected behavior.

Would it be possible to consider this adjustment as a performance enhancement for the package?

Thank you again for your continued support!

**Comment by huycozy** [comment]
>Would it be possible to consider this adjustment as a performance enhancement for the package?

Similar to what I shared above, even if this is a suggestion, it still needs an appropriate use case (reproduction sample) for better understanding.

**Comment by Mina-Zarif** [comment]
OK, Thanks for the feedback! I propose this as a performance enhancement for video_player, particularly for use cases involving rapid video switching. Here's a concrete example showing the performance impact:

```dart
import 'dart:developer';
import 'dart:io';

import 'package:file_picker/file_picker.dart';
import 'package:flutter/material.dart';
import 'package:video_player/video_player.dart';
import 'package:chewie/chewie.dart';

class VideoPlayerReproduction extends StatefulWidget {
  const VideoPlayerReproduction({super.key});

  @override
  State<VideoPlayerReproduction> createState() =>
      _VideoPlayerReproductionState();
}

class _VideoPlayerReproductionState extends State<VideoPlayerReproduction>
    with WidgetsBindingObserver {

  final PageController _pageController = PageController();
  final List<String> _videos = [
    "https://flutter.github.io/assets-for-api-docs/assets/videos/bee.mp4",
    "https://flutter.github.io/assets-for-api-docs/assets/videos/butterfly.mp4",
  ];

  VideoPlayerController? _videoController;
  ChewieController? _chewieController;
  int _currentIndex = 0;
  bool _isLoading = true;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
    _initializeVideo(_currentIndex);
  }

  Future<void> _initializeVideo(int index) async {
    setState(() => _isLoading = true);

    try {
      _videoController = VideoPlayerController.networkUrl(
        Uri.parse(_videos[index]),
        videoPlayerOptions: VideoPlayerOptions(
          mixWithOthers: true,
          allowBackgroundPlayback: false,
        ),
      );

      await _videoController!.initialize();

     
       _chewieController = ChewieController(
          videoPlayerController: _videoController!,
          autoPlay: true,
          looping: true,
          showControls: false,
       );

       setState(() => _isLoading = false);
      
    } catch (e) {
      log("Error initializing video: $e");
      setState(() => _isLoading = false);
    }
  }

  void _onPageChanged(int index) async {
    if (_currentIndex != index) {
      _currentIndex = index;

      // Dispose old controller before initializing new one
      await _videoController?.dispose();
      _chewieController?.dispose();

      // Initialize new video
      await _initializeVideo(index);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: PageView.builder(
        scrollDirection: Axis.vertical,
        controller: _pageController,
        onPageChanged: _onPageChanged,
        itemCount: _videos.length,
        itemBuilder: (context, index) {
          if (_isLoading || _chewieController == null) {
            return const Center(child: CircularProgressIndicator());
          }
          if (_currentIndex != index) {
            return const ColoredBox(color: Colors.black);
          }

          return Chewie(controller: _chewieController!);
        },
      ),
      floatingActionButton: FloatingActionButton(
        child: const Icon(
          Icons.add,
          size: 30,
        ),
        onPressed: () async {
          FilePickerResult? result =
              await FilePicker.platform.pickFiles(type: FileType.video);

          if (result != null) {
            File file = File(result.files.single.path!);
          } else {
            // User canceled the picker
          }
        },
      ),
    );
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    _videoController?.dispose();
    _chewieController?.dispose();
    _pageController.dispose();
    super.dispose();
  }
}
```
**Proposed Enhancement Benefits:**

**1- Smoother Transitions:**

Gracefully handle multiple initialization attempts
Prevent crashes during rapid scrolling
Better support for preloading strategies


**2- Resource Efficiency:**

Proper cleanup of video controllers
Better memory management during rapid switches
Reduced likelihood of memory leaks


**3- Better Error Recovery:**

Graceful fallback when initialization fails
No app crashes during edge cases
Better handling of network conditions

Would you like me to provide any additional performance metrics or test cases?

**Comment by huycozy** [comment]
Can you narrow the sample code above to a simpler one? `file_picker` is unrelated to this issue and `chewie` is a 3rd party package. Please check [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) again.

**Comment by Mina-Zarif** [comment]
Thank you for your time and assistance.

After further investigation, I found that the issue does not appear in the minimal code I created. This suggests the problem is specific to the configuration or setup in my original project and not a general bug in the video_player package.

I’ll continue debugging my project to identify the cause, but for now, it seems unrelated to the package itself. If I uncover anything that could impact other users, I’ll provide an update or reopen the issue.

Thanks again for your help!
